File: Longest Repeating Character Replacement - Leetcode 424/Longest Repeating Character Replacement - Leetcode 424.js
var characterReplacement = function(s, k) {
    const counts = new Array(26).fill(0);
    let l = 0, longest = 0, maxCount = 0;

    for (let r = 0; r < s.length; r++) {
        maxCount = Math.max(maxCount, ++counts[s.charCodeAt(r) - 65]);

        while ((r - l + 1) - maxCount > k) {
            counts[s.charCodeAt(l) - 65]--;
            l++;
        }

        longest = Math.max(longest, r - l + 1);
    }

    return longest;
};

File: Longest Repeating Character Replacement - Leetcode 424/Longest Repeating Character Replacement - Leetcode 424.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        longest = 0
        l = 0
        counts = [0] * 26

        for r in range(len(s)):
            counts[ord(s[r]) - 65] += 1

            while (r - l + 1) - max(counts) > k:
                counts[ord(s[l]) - 65] -= 1
                l += 1

            longest = max(longest, (r - l + 1))

        return longest

# Time Complexity: O(n)
# Space Complexity: O(1)
